pkg/ioutils/bytespipe.go

```diff
@@ -0,0 +1,79 @@
+package ioutils
+
+const maxCap = 1e6
+
+// BytesPipe is io.ReadWriter which works similary to pipe(queue).
+// All written data could be read only once. Also BytesPipe is allocating
+// and releasing new byte slices to adjust to current needs, so there won't be
+// overgrown buffer after highload peak.
+// BytesPipe isn't goroutine-safe, caller must synchronize it if needed.
+type BytesPipe struct {
+	buf      [][]byte // slice of byte-slices of buffered data
+	lastRead int      // index in the first slice to a read point
+	bufLen   int      // length of data buffered over the slices
+}
+
+// NewBytesPipe creates new BytesPipe, initialized by specified slice.
+// If buf is nil, then it will be initialized with slice which cap is 64.
+// buf will be adjusted in a way that len(buf) == 0, cap(buf) == cap(buf).
+func NewBytesPipe(buf []byte) *BytesPipe {
+	if cap(buf) == 0 {
+		buf = make([]byte, 0, 64)
+	}
+	return &BytesPipe{
+		buf: [][]byte{buf[:0]},
+	}
+}
+
+// Write writes p to BytesPipe.
+// It can allocate new []byte slices in a process of writing.
+func (bp *BytesPipe) Write(p []byte) (n int, err error) {
+	for {
+		b := bp.buf[len(bp.buf)-1]
+		n := copy(b[len(b):cap(b)], p)
+		bp.bufLen += n
+		bp.buf[len(bp.buf)-1] = b[:len(b)+n]
+		if len(p) > n { // more data: write to the next slice
+			p = p[n:]
+			nextCap := 2 * cap(bp.buf[len(bp.buf)-1])
+			if maxCap < nextCap {
+				nextCap = maxCap
+			}
+			bp.buf = append(bp.buf, make([]byte, 0, nextCap))
+			continue
+		}
+		break
+	}
+	return
+}
```

Comment:
Suggest:

                if len(p) == n {
                        break
                }

                // more data: write to the next slice
                p = p[n:]
                nextCap := 2 * cap(bp.buf[writeIndex])
                if maxCap < nextCap {
                        nextCap = maxCap
                }
                bp.buf = append(bp.buf, make([]byte, 0, nextCap))